File: FizzBuzzFP.hs
module FizzBuzzFP where

fizzBuzz :: Int -> String
fizzBuzz n
  | n < 1                            = error "n cannot be less than 1"
  | isDivisibleBy3 && isDivisibleBy5 = "Fizz Buzz"
  | isDivisibleBy3                   = "Fizz"
  | isDivisibleBy5                   = "Buzz"
  | otherwise                        = show n
  where
    isDivisibleBy3 = n `mod` 3 == 0
    isDivisibleBy5 = n `mod` 5 == 0

main :: IO ()
main = mapM_ (print . fizzBuzz) [1..30]

File: fizz_buzz_fp.clj
(ns fizz-buzz-fp)

(defn fizz-buzz [n]
  (let [divisible-by-3? (zero? (mod n 3))
        divisible-by-5? (zero? (mod n 5))]
    (cond
      (< n 1)               (throw (IllegalArgumentException. "n cannot be less than 1"))
      (and divisible-by-3?
           divisible-by-5?) "Fizz Buzz"
      divisible-by-3?       "Fizz"
      divisible-by-5?       "Buzz"
      :else                 (str n))))

;; pattern 1
(doall (map (comp println fizz-buzz) (range 1 (inc 30))))

;; pattern 2
(->> (range 1 (inc 30))
     (map (comp println fizz-buzz))
     doall)
